aes_cmac/aes.c

```diff
@@ -168,86 +168,143 @@ static NTSTATUS import_key(BCRYPT_ALG_HANDLE hAlg, BCRYPT_KEY_HANDLE *phKey,
   return !BCRYPT_SUCCESS(status);
 }
 
-#endif
-
-uint8_t aes_set_encrypt_key(uint8_t *key, uint16_t key_len, aes_context *ctx) {
-#ifdef _WIN32_BCRYPT
-  NTSTATUS status = STATUS_SUCCESS;
+#else
 
-  if (!BCRYPT_SUCCESS(status = init_ctx(ctx))) {
-    return 1;
+static const EVP_CIPHER *aes_ecb(uint16_t key_len) {
+  switch (key_len) {
+    case 16:
+      return EVP_aes_128_ecb();
+    case 24:
+      return EVP_aes_192_ecb();
+    case 32:
+      return EVP_aes_256_ecb();
+    default:
+      return NULL;
   }
+}
 
-  if (!BCRYPT_SUCCESS(status = import_key(ctx->hAlgCBC, &(ctx->hKeyCBC),
-                                          &(ctx->pbKeyCBCObj), ctx->cbKeyObj,
-                                          key, key_len))) {
-    return 1;
+static const EVP_CIPHER *aes_cbc(uint16_t key_len) {
+  switch (key_len) {
+    case 16:
+      return EVP_aes_128_cbc();
+    case 24:
+      return EVP_aes_192_cbc();
+    case 32:
+      return EVP_aes_256_cbc();
+    default:
+      return NULL;
   }
+}
 
-  if (!BCRYPT_SUCCESS(status = import_key(ctx->hAlgECB, &(ctx->hKeyECB),
-                                          &(ctx->pbKeyECBObj), ctx->cbKeyObj,
-                                          key, key_len))) {
-    return 1;
+static int aes_encrypt_ex(const EVP_CIPHER *cipher, uint8_t *in, uint8_t *out,
+                          uint16_t len, uint8_t *iv, aes_context *ctx) {
+  if (EVP_EncryptInit_ex(ctx->ctx, cipher, NULL, ctx->key, iv) != 1) {
+    return -1;
+  }
+  EVP_CIPHER_CTX_set_padding(ctx->ctx, 0);
```

Comment:
can we add error check for this call as well?

Reply:
Docs say it always returns 1

Reply:
And code agrees

Reply:
But I'll add checks just in case that changes